### PR TITLE
Add lock component documentation

### PR DESCRIPTION
**Pull request by kbickar**
## Description:
Add documentation for lock components

Update webserver js file to support locks.  Refactored webserver js file to have smaller file size

**Related issue (if applicable):** fixes <link to issue>

**Pull request in [esphome](https://github.com/esphome/esphome) with YAML changes (if applicable):** esphome/esphome#2958

## Checklist:

  - [X] Branch: `next` is for changes and new documentation that will go public with the next ESPHome release. Fixes, changes and adjustments for the current release should be created against `current`.
  - [X] Link added in `/index.rst` when creating new documents for new components or cookbook.

### CODE REVIEW

File: _static/webserver-v1.js [hunk 1]
```diff
@@ -2,25 +2,35 @@ const source = new EventSource("/events");
...
+    log.innerHTML += '<span class="' + klass + '">' + e.data.substr(7, e.data.length - 11) + "</span>\n";
 });
 
+actions = [
+    ["switch", ["toggle"]],
+    ["light", ["toggle"]],
+    ["cover", ["open", "close"]],
+    ["button", ["press"]],
+    ["lock", ["lock", "unlock", "open"]],
+    ];
```
Comment (jesserockz) [inline _static/webserver-v1.js]: You accidentally dropped fan from this list?

File: components/lock/index.rst [hunk 2]
```diff
@@ -0,0 +1,168 @@
+Lock Component
+================
+
+.. seo::
+    :description: Instructions for setting up generic lockss in ESPHome.
```
Comment (jesserockz) [inline components/lock/index.rst]:
```suggestion
    :description: Instructions for setting up generic locks in ESPHome.
```

File: components/lock/index.rst [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+- **internal** (*Optional*, boolean): Mark this component as internal. Internal components will
+  not be exposed to the frontend (like Home Assistant). Only specifying an ``id`` without
+  a ``name`` will implicitly set this to true.
+- **on_lock** (*Optional*, :ref:`Action <config-action>`): An automation to perform
+  when the lock is locked. See :ref:`lock-on_lock_unlock_trigger`.
+- **on_unlock** (*Optional*, :ref:`Action <config-action>`): An automation to perform
+  when the lock is unlocked. See :ref:`lock-on_lock_unlock_trigger`..
+- **disabled_by_default** (*Optional*, boolean): If true, then this entity should not be added to any client's frontend,
+  (usually Home Assistant) without the user manually enabling it (via the Home Assistant UI).
+  Requires Home Assistant 2021.9 or newer. Defaults to ``false``.
```
Comment (jesserockz) [inline components/lock/index.rst]: Same for entity category below as well, put a `note` at the top of lock similar to number for HA version:
https://github.com/esphome/esphome-docs/blob/425e0bc38d35b7945eb086d2403c1a076f89b764/components/number/index.rst?plain=1#L12-L14

```suggestion
  Defaults to ``false``.
```

File: components/lock/index.rst [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    on_...:
+      then:
+        - lock.unlock: deadbolt_1
+
+.. _lock-open_action:
+
+``lock.open`` Action
+************************
+
+This action open (e.g. unlatch) a lock with the given ID off when executed.
```
Comment (jesserockz) [inline components/lock/index.rst]:
```suggestion
This action opens (e.g. unlatch) a lock with the given ID off when executed.
```

File: components/lock/index.rst [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+    on_...:
+      then:
+        - lock.lock: deadbolt_1
+
+.. _lock-unlock_action:
+
+``lock.unlock`` Action
+**************************
+
+This action unlock a lock with the given ID off when executed.
```
Comment (jesserockz) [inline components/lock/index.rst]:
```suggestion
This action unlocks a lock with the given ID off when executed.
```

File: components/lock/template.rst [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    # Example configuration entry
+    lock:
+      - platform: template
+        name: "Template Lock"
+        lambda: |-
+          if (id(some_binary_sensor).state) {
+            return true;
+          } else {
+            return false;
+          }
```
Comment (jesserockz) [inline components/lock/template.rst]: This should probably use the enums instead of true/false

File: components/lock/template.rst [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+      # Templated
+      - lock.template.publish:
+          id: template_lock1
+          state: !lambda 'return true;'
+
+Configuration options:
+
+- **id** (**Required**, :ref:`config-id`): The ID of the template lock.
+- **state** (**Required**, boolean, :ref:`templatable <config-templatable>`):
+  The state to publish.
```
Comment (jesserockz) [inline components/lock/template.rst]: State here should match the enums used in config validation.

File: components/lock/template.rst [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+.. note::
+
+    This action can also be written in lambdas, the parameter of the `public_state` method denotes if
+    the lock is currently locked or unlocked:
```
Comment (jesserockz) [inline components/lock/template.rst]:
```suggestion
    This action can also be written in lambdas, the parameter of the `publish_state` method denotes the state the lock should become:
```

File: components/lock/template.rst [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+
+    .. code-block:: cpp
+
+        id(template_lock1).publish_state(false);
```
Comment (jesserockz) [inline components/lock/template.rst]:
```suggestion
        id(template_lock1).publish_state(lock::LOCK_STATE_LOCKED);
```

File: web-api/index.rst [hunk 10]
```diff
@@ -249,3 +249,13 @@ can be used:
...
+
+Lock
+******
+
+Locks have the exact same properties as binary sensors in the state reporting aspect, but they
+additionally support setting states with the ``lock``, ``unlock`` and ``open`` methods.
+
+Each of these is quite self explanatory. Creating a POST request to ``/lock/front_door/unlock``
+would for example result in the component called "Front Door" to be unlocked. The server will respond
+with a 200 OK HTTP return code if the call succeeded.
```
Comment (jesserockz) [inline web-api/index.rst]: This needs to be updated to include the other potential states.

**Review by jesserockz — CHANGES_REQUESTED** [review]

**Comment by kbickar** [comment]
Ok, I think I broke it....https://github.com/esphome/esphome-docs/pull/1867

**Pull request closed without merging**